### ISSUE TITLE
Multi-Module support

### ISSUE THREAD

**Issue by ABHAY0O7** [castor-software/depclean#68]
Hey! @cesarsotovalero,
As mentioned in the CASTOR's GSoC idealist for DepClean project that there is a need for the support of analysis of the multi-module projects.
I want to ask that can you describe this feature a little bit more? 
As there is no existing issue or PR regarding this feature so I am unable to getting what is the current situation of this issue (feature) and what is needs to be implemented.

Thank you

**Comment by cesarsotovalero** [comment]
Hi @ABHAY0O7,

### Problem description

DepClean currently analyzes a single module of a Maven project. For example, let's say that a project has three modules `A`, `B`, and `C`. You can run DepClean to get the usage status of dependency of, say, project `A`. Suppose that you found that dependency `x` in module `A` is not used. So, DepClean reports this dependency as "potentially unused". However, suppose that module `B` depends on module `A` and it uses dependency `x`. Then, if the developer removes the dependency `x` from `A` the project will not build because module `B` complains about the missing dependency `x`.

### Objective

We want to make DepClean able to analyze multi-module projects and provide a report that takes into account the dependency between the modules. To do so, we need to add support for the analysis and aggregation of the individual report using by leveraging the [reactor Maven component](https://maven.apache.org/guides/mini/guide-multiple-modules.html). This way, we will obtain a much more accurate report and provide the ability to analyze complex projects.

**Note:** If you run DepClean today on the parent of a multi-module project, then DepClean will be executed on each module individually. Thus providing individual reports for each module without taking into account the case described previously.

**Comment by ABHAY0O7** [comment]
Hello @cesarsotovalero 
I came up with an idea for this.
<b>Idea : </b>After generating a depClean analysis report for all the modules (say `w, x, y, z`), we can compare two sets 
`usedUndeclaredArtifacts` of `w` with `unusedDeclaredArtifacts` of all other modules `x, y, z`.
`usedUndeclaredArtifacts` of `x` with `unusedDeclaredArtifacts` of all other modules `w, y, z`.
Similarly for others.

At the time of comparing if any element of both the set matches then we store that element in a new set.(common for all modules)
And In the end if set is not empty then we can put it in a different section in the analysis report with name (not decided now) before this part
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] MavenReactor 1.0.0-SNAPSHOT ........................ SUCCESS [  1.281 s]
[INFO] MavenReactorModule1 1.0.1-SNAPSHOT ................. SUCCESS [  3.163 s]
[INFO] MavenReactorModule2 1.0.1-SNAPSHOT ................. SUCCESS [  1.660 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  6.245 s
[INFO] Finished at: 2021-04-11T18:46:39+05:30
[INFO] ------------------------------------------------------------------------

**Comment by ABHAY0O7** [comment]
Also, @cesarsotovalero,
Can you please tell me that depClean follows which of the following mechanism in a multi-module(say A, B, C, D) project?
a) Analyze module A and output it to the console, then analyzes module B and output it to the console, then for C and D. or,
b) Analyze module A and store the report, then analyze module B and store the report, then C and then D. And in the end output all the reports to the console.

Waiting for your reply...

**Comment by cesarsotovalero** [comment]
> Hey! @cesarsotovalero,
> Please, can you also tell me that depClean follows which of the following mechanism in a multi-module(say A, B, C, D) project?
> a) Analyze module A and output it to the console, then analyzes module B and output it to the console, then for C and D. or,
> b) Analyze module A and store the report, then analyze module B and store the report, then C and then D. And in the end output all the reports to the console.
> 
> Waiting for your response...

Hi @ABHAY0O7. At the moment, DepClean follows the approach that you described in a). Each module is analyzed one by one, separately.

**Comment by cesarsotovalero** [comment]
> Hello @cesarsotovalero
> I came up with an idea for this.
> Idea : After generating a depClean analysis report for all the modules (say `w, x, y, z`), we can compare two sets
> `usedUndeclaredArtifacts` of `w` with `unusedDeclaredArtifacts` of all other modules `x, y, z`.
> `usedUndeclaredArtifacts` of `x` with `unusedDeclaredArtifacts` of all other modules `w, y, z`.
> Similarly for others.
> 
> At the time of comparing if any element of both the set matches then we store that element in a new set `dependenciesOfDependentModules`. (common for all modules)
> And In the end, if this 👆 set is not empty then we can print it in a different section in the analysis report with the description `Dependencies shared between modules are` and then printing the complete set.
> 
> Also, we can mention the modules between whom the dependency is shared.
> 
> Thank you

Hi @ABHAY0O7,
What you propose is actually a meta-report that summarizes the analysis of the individual modules at the end of the multi-module analysis. This is nice to have, and also a good starting point for resolving this task 👍 . 

Notice that the proposed approach doesn't consider interdependencies between modules. To properly handle a multi-module project, we need to have information on all the types used by each module and cross-ref them. Then, it is necessary to analyze this bulk of data in order to give a proper summary report.

**Comment by ABHAY0O7** [comment]
Hello @cesarsotovalero
> Notice that the proposed approach doesn't consider interdependencies between modules. To properly handle a multi-module project, we need to have information on all the types used by each module and cross-ref them. Then, it is necessary to analyze this bulk of data in order to give a proper summary report.

I understand that we need to consider interdependencies between modules. But, it is impossible with the current situation of the maven plugin. 
So, I think the first step to support MMA is to change the current situation of the maven plugin from this

> a) Analyze module A and output it to the console, then analyzes module B and output it to the console, then for C and D. or,

to this 

> b) Analyze module A and store the report, then analyze module B and store the report, then C and then D. And in the end output all the reports to the console.

And only after achieving this, we can think of an approach to consider interdependencies between modules.

Thank you
